### PR TITLE
Add Sentry error reporting

### PR DESCRIPTION
**Pull request by poquirion**
Implement #1475

### CODE REVIEW

File: scripts/sct_utils.py [hunk 1]
```diff
@@ -137,6 +189,32 @@ def colors(cls):
         return [v for k, v in cls.__dict__.items() if not k.startswith("_") and k is not "colors"]
 
 
+class Config:
+    """
+    Config value use all around the sct
+    """
+    _sentry_config = None
+    _log_dsn = "uggcf://14p24045rq8s4nno97qp86n3o034nn5r:0226q799rss8450qn38oro87rsop904r@fragel.vb/232808"
```
Comment (peristeri) [inline scripts/sct_utils.py]: I don't agree with having the `_log_dsn` key available to all users. This should be used for internal use.

I would make this an environment variable instead, set it manually.

**Comment by zougloub** [comment]
I will wait to see the installation part, and a configuration.
Not sure about what the rot13 is achieving.

**Comment by poquirion** [comment]
...'rot13', its a pretty lame way hide a string real meaning.

**Comment by poquirion** [comment]
Following: https://github.com/getsentry/raven-python/issues/569, we need to use the public key.

File: install_sct [hunk 2]
```diff
@@ -258,10 +258,10 @@ fi
 # Questions
 echo
 cat <<EOF
-The SCT teams is constantly improving the toolbox and uses request and
-feedback to improve its user experience and fix bugs. We also have in
-place a system to receive anonymous crash report and errors from our
-users. If you want to participate and help the SCT improvement this way
+The SCT team is constantly improving the toolbox and uses request and
```
Comment (jcohenadad) [inline install_sct]: i would rather go with "yes" by default

**Comment by poquirion** [comment]
We will go for yes by default, but we still need a way to batch install the sct without sentry.

The sct batch install trick is usually `yes | install_sct` could become `yes | NO_SENTRY=y install_sct`

**Comment by poquirion** [comment]
As soon as the file for the sentry_dsn.txt file is moved to the right place, we should be able to merge to master.

**Comment by jcohenadad** [comment]
@poquirion i would make the variable name more intuitive for the user + code reader, i.e., use the affirmative form instead of the negative form. E.g.: `USE_SENTRY`, instead of `NO_SENTRY`. 
It is cognitively easier to say yes to something we want (or no to something we don't want), versus yes to something we don't want and no-no to something we want.

**Comment by poquirion** [comment]
@jcohenadad 
no prob.

**Comment by poquirion** [comment]
The

**Comment by poquirion** [comment]
The path on the wiki is updated. All seems ready for the PR

**Comment by poquirion** [comment]
@zougloub can you have a look?

File: scripts/sct_config.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                         cwd=__sct_dir__)
+    output, _ = p.communicate()
+    status = p.returncode
+    if status == 0:
+        return output
+    return 'unknown'
+
+
+# Basic sct config
+__sct_dir__ = os.getenv("SCT_DIR", os.path.dirname(os.path.realpath(__file__)).rstrip("scripts"))
```
Comment (zougloub) [inline scripts/sct_config.py]: `.rstrip(scripts) -> os.path.dirname()`

File: scripts/sct_config.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+__data_dir__ = os.getenv("SCT_DATA_DIR", "{}/data".format(__sct_dir__))
```
Comment (zougloub) [inline scripts/sct_config.py]: `%s/%s -> os.path.join(%s, %s)`

File: scripts/sct_config.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+# Be careful no to change the order commit, branch and then version!
```
Comment (zougloub) [inline scripts/sct_config.py]: Can we find a way to the scary comment (eg. by having a function that returns all in one shot)?

File: scripts/sct_utils.py [hunk 6]
```diff
@@ -58,6 +82,56 @@ def start_stream_logger():
...
+    """ Adds sentry log handler to the logger
+
+    :return: the sentry handler
+    """
+    from raven.handlers.logging import SentryHandler
+
+    sh = SentryHandler(client=client, level=sct_config.__report_log_level__)
+    fmt = ("[%(asctime)s][%(levelname)s] %(filename)s: %(lineno)d | "
+            "%(message)s")
+    formater = logging.Formatter(fmt=fmt, datefmt="%H:%M:%S")
```
Comment (zougloub) [inline scripts/sct_utils.py]: You fixed a "formating" -> "formatting" typo above

File: scripts/sct_check_dependencies.py [hunk 7]
```diff
@@ -17,17 +17,21 @@
...
 
+import commands
+import importlib
+import io
+import os
+import platform
+import sys
 
-import sys, io, os, commands, platform, importlib
+import sct_config
```
Comment (zougloub) [inline scripts/sct_check_dependencies.py]: Not really sure about the utility of `sct_config` given that most of the code inside was already in `sct_utils`.

**Comment by zougloub** [comment]
@poquirion sorry I hadn't seen your request to review. Here's some feedback.

**Comment by poquirion** [comment]
The reason to have a config file is to centralize that type of information I thought it was somewhat lost in sct_util. This way everybody knows that it should get the version, data dir, etc information from there instead of recomputing it in the code like it is done in many places.

File: install_sct [hunk 8]
```diff
@@ -36,6 +36,7 @@ SCRIPT_DIR="scripts"
 DATA_DIR="data"
 PYTHON_DIR="python"
 BIN_DIR="bin"
+STATS_DNS="https://www.neuro.polymtl.ca/_media/downloads/stats_dns.txt"
```
Comment (jcohenadad) [inline install_sct]: shouldn't we move that to SCT Github's repos instead?

**Comment by jcohenadad** [comment]
I've created a NeuroPoly account on sentry and updated the DNS [there](https://www.neuro.polymtl.ca/_media/downloads/stats_dns.txt).

**Comment by jcohenadad** [comment]
to test this PR, i've added `sct.printv('test', type='error')` at the beginning of `sct_maths`. When running the command, it tries to upload the error to Sentry, but does not succeed in doing so. Any idea why?
~~~
sct_maths -i t2s.nii.gz -add 1 -o test.nii.gz

--
Spinal Cord Toolbox (poq_issue_1475_sentry/61974aa5ceeb6b7e282fb1f623b31341bcfd026f)
Running /Users/julien/code/sct-sentry/scripts/sct_maths.py -i t2s.nii.gz -add 1 -o test.nii.gz
Configuring sentry report
sentry is set!
test

/Users/julien/code/sct-sentry/scripts/sct_maths.pyNone

Traceback (most recent call last):
  File "/Users/julien/code/sct-sentry/scripts/sct_maths.py", line 724, in <module>
    main()
  File "/Users/julien/code/sct-sentry/scripts/sct_maths.py", line 188, in main
    sct.printv('test', type='error')
  File "/Users/julien/code/sct-sentry/scripts/sct_utils.py", line 1021, in printv
    raise RunError('raise in printv, read log above for more info')
sct_utils.RunError: raise in printv, read log above for more info
Sentry is attempting to send 2 pending error messages
Waiting up to 10 seconds
Press Ctrl-C to quit
~~~

**Comment by perone** [comment]
I wasn't able to test with the neuropoly DSN because I don't have the access to it on Sentry website, but I just changed to a DSN that I created to test.

I added the following in the `sct_maths` script (after the call to `sct.init_sct()`):

    sct.printv('test', type='error')

And it worked, however with some issues:

1) It created to registers of error in Sentry
image

2) There was a warning about old version
image

3) I wasn't able to understand the message, it seems that it is capturing color ascii codes and sending to Sentry
image

4) It mentions an error while processing the event
image

@jcohenadad I think that it is also working for you because I have the same message that you got on your console. Try to check if don't have any filter hiding the error on the Sentry website. Try to check again. Also, try to confirm if the DSN is the correct one that you have on your neuropoly Sentry, you can check that by doing:

    import os
    print(os.environ["SENTRY_DSN"])

before the call to the `sct.init_sct()` function. This is the message I got when it worked:

image

File: scripts/sct_utils.py [hunk 9]
```diff
@@ -13,7 +13,22 @@
...
+import subprocess
+import sys
+import tempfile
+import time
+
+import sct_config
+
+if os.getenv('SENTRY_DSN', None):
+    # do no import if not needed
+    import raven
```
Comment (jcohenadad) [inline scripts/sct_utils.py]: @poquirion when I run this line (on OSX_10.13), the variable `SENTRY_DSN` is still `None` afterwards, therefore I don't see the point in adding this line

Comment (jcohenadad) [reply scripts/sct_utils.py]: ok nevermind-- i figured that out and in the meantime explicited the line of code here: https://github.com/neuropoly/spinalcordtoolbox/pull/1544/commits/a3daba57ef41cccff1f10424a0645288258ccc75

**Comment by jcohenadad** [comment]
@poquirion could you please address [Christian's comments](https://github.com/neuropoly/spinalcordtoolbox/pull/1544#issuecomment-375993885) 1-4, which I also experienced.

**Comment by perone** [comment]
@jcohenadad I think that the best way to debug in PyCharm is to add the environment variable on the debugging as you said. Another alternative is to manually set in the code by adding:

    os.environ["SENTRY_DSN"] = '[desired dsn]'

**Comment by poquirion** [comment]
For the sct setting in pycharm, you can source the sct_launcher in a terminal an then launch pycharm in that same terminal.

**Comment by jcohenadad** [comment]
@poquirion there are still some conflicts, will you do it or should someone else do it?

**Comment by zougloub** [comment]
Rebased and squashed the changes into one commit.

**Comment by jcohenadad** [comment]
We should let users opt out at anytime: https://github.com/neuropoly/spinalcordtoolbox/issues/1708

**Pull request merged**